File: flow_sdk/claude_hook_events/field_extractors.py
"""Pure field extractor functions for HookEventData.

Phase 9: HookEventData carries `process_entry` (typed) + `extra` (raw payload
spillover) instead of 30+ flat optional fields. These helpers read from both
seams transparently.
"""

from __future__ import annotations

from typing import Any, Optional

from flow_sdk.claude_hook_events.hook_event_data import HookEventData


def get_trigger_log_dock_pointer(
    hook_entry_id: Optional[str],
) -> Optional[dict]:
    """Compute a trigger log dock pointer from a hook entry ID."""
    if not hook_entry_id:
        return None
    return {"ref": hook_entry_id, "options": {}}


def _crop_text(text: str, max_words: int = 5) -> str:
    words = text.split()
    if len(words) <= max_words:
        return " ".join(words)
    return " ".join(words[:max_words]) + "..."


def _extra(hook_data: HookEventData, key: str) -> Any:
    """Read a variant-specific field that used to be a flat HookEventData
    attribute and now lives in ``extra``."""
    return (hook_data.extra or {}).get(key)


def extract_cwd(hook_data: HookEventData) -> str | None:
    return hook_data.cwd


def extract_session_id(hook_data: HookEventData) -> str | None:
    return hook_data.session_id


def extract_session_id_from_dict(hook_data: dict) -> str | None:
    """Extract session_id from a raw webhook hook_data dict.

    Walks the same precedence chain the UI hook uses:
      1. hook_data.raw_hook_data.session_id
      2. hook_data.session_id
      3. hook_data.event.context.session_id
    """
    if not isinstance(hook_data, dict):
        return None
    raw = hook_data.get("raw_hook_data") if isinstance(hook_data.get("raw_hook_data"), dict) else None
    if raw and raw.get("session_id"):
        return str(raw["session_id"]) or None
    if hook_data.get("session_id"):
        return str(hook_data["session_id"]) or None
    event = hook_data.get("event") if isinstance(hook_data.get("event"), dict) else None
    context = event.get("context") if event and isinstance(event.get("context"), dict) else None
    if context and context.get("session_id"):
        return str(context["session_id"]) or None
    return None


def get_tool_name(hook_data: HookEventData) -> str | None:
    """Read tool_name from process_entry (typed) or extra (raw)."""
    if hook_data.process_entry:
        te = hook_data.process_entry.get("transcript_entry") or {}
        if te.get("tool_name"):
            return str(te["tool_name"])
    return _extra(hook_data, "tool_name")


def get_event_summary_line(hook_data: HookEventData) -> str:
    """One-line plain-text summary for a hook event."""
    event_name = hook_data.hook_event_name

    tool_name = get_tool_name(hook_data)
    tool_input = _extra(hook_data, "tool_input")
    if hook_data.process_entry and not tool_input:
        te = hook_data.process_entry.get("transcript_entry") or {}
        tool_input = te.get("tool_input")

    if tool_name:
        if isinstance(tool_input, dict) and tool_input:
            for k in ("command", "file_path", "pattern", "url", "query", "description", "prompt"):
                v = tool_input.get(k)
                if isinstance(v, str) and v:
                    return f"{tool_name}: {_crop_text(v)}"
        return tool_name

    for k in ("message", "prompt", "notification_type", "agent_type", "reason",
              "task_subject", "teammate_name", "source", "name", "worktree_path"):
        v = _extra(hook_data, k)
        if isinstance(v, str) and v:
            return f"{event_name}: {_crop_text(v)}"

    return event_name or ""

File: ts_sdk/src/claude_hook_events/field-extractors.ts
/**
 * Pure field extractor functions for HookEventData.
 */

import { HookEventType } from './event-types';
import type { HookEventData } from './hook-event-data';

// ---------------------------------------------------------------------------
// Transcript path helpers
// ---------------------------------------------------------------------------

/**
 * Parse a Claude transcript path into its project-encoded-name and session ID.
 * Handles paths like: /home/user/.claude/projects/-My-Project/uuid.jsonl
 * Returns null if the path doesn't match the expected format.
 */
export function parseTranscriptPath(path: string): { projectEncodedName: string; sessionId: string } | null {
  const projectsIdx = path.indexOf('.claude/projects/');
  if (projectsIdx < 0) return null;
  const remainder = path.substring(projectsIdx + '.claude/projects/'.length);
  const slashIdx = remainder.indexOf('/');
  if (slashIdx < 0) return null;
  return {
    projectEncodedName: remainder.substring(0, slashIdx),
    sessionId: remainder.substring(slashIdx + 1).replace(/\.jsonl$/, ''),
  };
}

/**
 * Compute a trigger log dock pointer from a hook entry ID.
 *
 * The hook entry ID is the trigger entity ID — the same value TriggerLogViewer
 * uses as `triggerId` to fetch `GET /trigger/{id}/log`.
 *
 * Returns null when no hook entry ID is present.
 */
export function getTriggerLogDockPointer(
  hookEntryId: string | null | undefined,
): { ref: string; options: Record<string, string> } | null {
  if (!hookEntryId) return null;
  return { ref: hookEntryId, options: {} };
}

/**
 * Compute a transcript dock pointer from hook event data.
 *
 * Phase 9: prefers `session_id` directly (single-segment ref form matching
 * `process.transcriptDockPointer`'s new shape). Falls back to parsing the
 * legacy `transcript_path` for old payloads that don't carry session_id.
 *
 * Returns null when:
 * - The event is SessionStart (transcript file not yet created on disk)
 * - Neither session_id nor a parseable transcript_path is present
 */
export function getTranscriptDockPointer(
  hookData: HookEventData,
  timestamp?: string,
): { ref: string; options: Record<string, string> } | null {
  if (hookData.hook_event_name === HookEventType.SESSION_START) return null;
  const options: Record<string, string> = {};
  if (timestamp) options.ts = timestamp;
  if (hookData.session_id) {
    return { ref: hookData.session_id, options };
  }
  if (hookData.transcript_path) {
    const parsed = parseTranscriptPath(hookData.transcript_path);
    if (parsed) {
      return { ref: parsed.sessionId, options };
    }
  }
  return null;
}

function cropText(text: string, maxWords = 5): string {
  const words = text.split(/\s+/).filter(Boolean);
  if (words.length <= maxWords) return words.join(' ');
  return words.slice(0, maxWords).join(' ') + '...';
}

/** Extract the working directory from hook event data. */
export function extractCwd(hookData: HookEventData): string | null {
  return hookData.cwd ?? null;
}

/** Extract the session ID from hook event data. */
export function extractSessionId(hookData: HookEventData): string | null {
  return hookData.session_id ?? null;
}

/** Extract the tool name from hook event data. */
export function getToolName(hookData: HookEventData): string | null {
  return hookData.tool_name ?? null;
}

/**
 * Get a single-line plain-text summary for a hook event.
 *
 * Tool events: "ToolName: key-input-detail"
 * Non-tool events: "EventName: detail"
 */
export function getEventSummaryLine(hookData: HookEventData): string {
  const toolName = hookData.tool_name;
  const toolInput = hookData.tool_input;
  const eventName = hookData.hook_event_name;

  if (toolName) {
    if (toolInput) {
      const keyField =
        toolInput['command'] ||
        toolInput['file_path'] ||
        toolInput['pattern'] ||
        toolInput['url'] ||
        toolInput['query'] ||
        toolInput['description'] ||
        toolInput['prompt'] ||
        '';
      if (typeof keyField === 'string' && keyField) {
        return `${toolName}: ${cropText(keyField)}`;
      }
    }
    return toolName;
  }

  if (hookData.message) return `${eventName}: ${cropText(hookData.message)}`;
  if (hookData.prompt) return `${eventName}: ${cropText(hookData.prompt)}`;
  if (hookData.notification_type) return `${eventName}: ${hookData.notification_type}`;
  if (hookData.agent_type) return `${eventName}: ${hookData.agent_type}`;
  if (hookData.reason) return `${eventName}: ${hookData.reason}`;
  if (hookData.task_subject) return `${eventName}: ${cropText(hookData.task_subject)}`;
  if (hookData.teammate_name) return `${eventName}: ${hookData.teammate_name}`;
  if (hookData.source) return `${eventName}: ${hookData.source}`;
  if (hookData.name) return `${eventName}: ${hookData.name}`;
  if (hookData.worktree_path) return `${eventName}: ${cropText(hookData.worktree_path)}`;

  return eventName || '';
}
